File: leetcode/787.cheapest-flights-within-k-stops.js
const { MinPriorityQueue } = require('@datastructures-js/priority-queue');

/*
 * @lc app=leetcode id=787 lang=javascript
 *
 * [787] Cheapest Flights Within K Stops
 */

// @lc code=start
/**
 * @param {number} n
 * @param {number[][]} flights
 * @param {number} src
 * @param {number} dst
 * @param {number} k
 * @return {number}
 */
var findCheapestPrice = function (n, flights, src, dst, k) {
    // prepare graphs
    const fullGraph = []
    const visited = []
    const cutGraph = []

    for (let i = 0; i < n; i++) {
        fullGraph.push([])
        cutGraph.push([])
        visited.push(false)
    }

    for (let i = 0; i < flights.length; i++) {
        const [from, to, price] = flights[i]

        fullGraph[from].push([to, price])
    }

    // console.log(fullGraph)
    // console.log(cutGraph)

    // cut the graph by K
    const stack = []

    stack.push({ node: src, stops: 0 })
    visited[src] = true

    while (stack.length) {
        const { node, stops } = stack.pop()
        const pricedNeighbours = fullGraph[node]

        // console.log(pricedNeighbours, node, stops)

        for (let i = 0; i < pricedNeighbours.length; i++) {
            const [neighbour, price] = pricedNeighbours[i]

            if (!visited[neighbour] && stops < k || neighbour === dst) {
                stack.push({ node: neighbour, stops: stops + 1 })
                visited[neighbour] = true
                cutGraph[node].push([neighbour, price])
            }
        }
    }

    // console.log(visited)
    // console.log(cutGraph)

    // traverce cutGraph starting from src's neighbours
    const minHeap = new MinPriorityQueue(([node, price]) => price);

    for (let i = 0; i < cutGraph[src].length; i++) {
        const [neigbour, price] = cutGraph[src][i]

        minHeap.push([neigbour, price])
    }

    let sum = 0

    while (!minHeap.isEmpty()) {
        const [node, price] = minHeap.pop()

        sum = sum + price

        if (node === dst) {
            return sum
        }

        const pricedNeighbours = cutGraph[node]

        for (let i = 0; i < pricedNeighbours.length; i++) {
            const [neighbour, price] = pricedNeighbours[i]

            minHeap.push([neighbour, price])
        }
    }

    return -1
};
// @lc code=end

{
    const flights = [[0, 1, 100], [1, 2, 100], [2, 0, 100], [1, 3, 600], [2, 3, 200]]
    const src = 0
    const dst = 3
    const k = 1
    const n = 4;

    const expected = 700;
    const result = findCheapestPrice(n, flights, src, dst, k);

    console.log(expected === result, result);
    console.log("------------------");
}

{
    const flights = [[0, 1, 100], [1, 2, 100], [2, 0, 100], [1, 3, 600], [2, 3, 200]]
    const src = 0
    const dst = 3
    const k = 2
    const n = 4;

    const expected = 400;
    const result = findCheapestPrice(n, flights, src, dst, k);

    console.log(expected === result, result);
    console.log("------------------");
}

{
    const n = 3;
    const flights = [[0, 1, 100], [1, 2, 100], [0, 2, 500]]
    const src = 0
    const dst = 2
    const k = 1

    const expected = 200;
    const result = findCheapestPrice(n, flights, src, dst, k);

    console.log(expected === result, result);
    console.log("------------------");
}

{
    const n = 3;
    const flights = [[0, 1, 100], [1, 2, 100], [0, 2, 500]]
    const src = 0
    const dst = 2
    const k = 0

    const expected = 500;
    const result = findCheapestPrice(n, flights, src, dst, k);

    console.log(expected === result, result);
    console.log("------------------");
}

File: leetcode/787.cheapest-flights-within-k-stops.py
from typing import List
from queue import PriorityQueue, Queue
import math



'''
787. Cheapest Flights Within K Stops

There are n cities connected by some number of flights.
You are given an array flights where flights[i] = [fromi, toi, pricei]
indicates that there is a flight from city fromi to city toi with cost pricei.

You are also given three integers src, dst, and k, return the cheapest price from src to dst with at most k stops.
If there is no such route, return -1.
'''

class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        graph = [[] for _ in range(n)]

        for u, v, price in flights:
            graph[u].append((v, price))

        # print(graph)

        k += 1

        queue = PriorityQueue()

        queue.put((0, src, 0))

        minNodeEdges = [float("inf")] * n

        while not queue.empty():
            current_distance, current_node, current_edges = queue.get()

            if current_node == dst: 
                return current_distance

            if current_edges > minNodeEdges[current_node]:
                continue

            minNodeEdges[current_node] = current_edges

            for neighbour, weight in graph[current_node]:
                nextDistance = current_distance + weight
                nextEdges = current_edges + 1

                if nextEdges <= k:
                    queue.put((nextDistance, neighbour, nextEdges))

        return -1


testCases = [
    {
        "flights": [
            [0,1,100],
            [1,2,100],
            [2,0,100],
            [1,3,600],
            [2,3,200]
        ],
        "n": 4,
        "src": 0,
        "dst": 3,
        "k": 1,
        "expected": 700
    },
    {
        "flights": [
            [0,1,1],
            [0,2,5],
            [1,2,1],
            [2,3,1]
        ],
        "n": 4,
        "src": 0,
        "dst": 3,
        "k": 1,
        "expected": 6
    },
    {
        "flights": [
            [0,1,1],
            [0,2,5],
            [1,2,1],
            [2,3,1],
            [1,3,4],
        ],
        "n": 4,
        "src": 0,
        "dst": 3,
        "k": 1,
        "expected": 5
    },
    {
        "flights": [
            [0,1,100],
            [1,2,100],
            [0,2,500]
        ],
        "n": 3,
        "src": 0,
        "dst": 2,
        "k": 1,
        "expected": 200
    },
    {
        "flights": [
            [0,3,3],
            [3,4,3],
            [4,1,3],
            [0,5,1],
            [5,1,100],
            [0,6,2],
            [6,1,100],
            [0,7,1],
            [7,8,1],
            [8,9,1],
            [9,1,1],
            [1,10,1],
            [10,2,1],
            [1,2,100]
        ],
        "n": 11,
        "src": 0,
        "dst": 2,
        "k": 4,
        "expected": 11
    },
]


for testCase in testCases:
    print('')

    flights = testCase["flights"]
    n = testCase["n"]
    k = testCase["k"]
    src = testCase["src"]
    dst = testCase["dst"]
    expected = testCase["expected"]

    s = Solution()

    result = s.findCheapestPrice(n, flights, src, dst, k)
    print(n, flights, src, dst, k, result)
    assert result == expected, f"result {result} should be expected: {expected}"
